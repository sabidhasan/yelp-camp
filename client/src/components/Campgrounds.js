import React from 'react'
import PropTypes from 'prop-types'
import CampgroundTile from './CampgroundTile'
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';

class Campgrounds extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      spin: false,
      campgrounds: []
    };

    this.newRandoms = this.newRandoms.bind(this);
  }

  static contextTypes = {
    startLoad: PropTypes.func,
    finishLoad: PropTypes.func
  }

  newRandoms() {
    // Reset the state, then add 10 new campgrounds to it
    // this.context.startLoad(this.constructor.name)
    fetch('/campground?random=true')
      .then(res => res.json())
      .then(cg => this.setState({ campgrounds: cg }));
      // .then(cg => this.setState({ campgrounds: cg }, () => this.context.finishLoad(this.constructor.name)))
  }

  componentDidMount() {
    this.spinner.addEventListener('click', () => this.setState({spin: true}));
    this.spinner.addEventListener('animationend', () => this.setState({spin: false}));
    this.newRandoms();
  }

  render() {
    return (
      <ul className='Campgrounds'>
        <div className='Campgrounds__header flex-center'>
          <h1 className='Campgrounds__header-title'>Some Awesome Campgrounds</h1>
          <i
            className={`Campgrounds__random fas fa-sync-alt ${this.state.spin ? 'Campgrounds__random--spin' : ''}`}
            onClick={this.newRandoms}
            ref={spinner => this.spinner = spinner}
            ></i>
        </div>
        {this.state.campgrounds.map(val => {
          const rating = val.comments.reduce((acc, val) => acc + val.rating, 0) / val.comments.length || 0
          return (
            <li key={val['_id']} className='CampgroundTile'>
              <CampgroundTile
                id={val.id}
                name={val.name}
                image={val.image}
                region={val.region}
                province={val.province}
                rating={rating}
                ratingCount={val.comments.length}
              />
            </li>
          )})
        }
      </ul>
    )
  }
}

export default Campgrounds
