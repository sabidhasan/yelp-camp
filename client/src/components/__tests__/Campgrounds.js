import React from 'react'
import toJson from 'enzyme-to-json'
import { MemoryRouter } from 'react-router-dom'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Campgrounds from '../Campgrounds'

configure({ adapter: new Adapter() });

describe('<Campgrounds />', () => {
  let component;
  const newRandoms = () => {
    return new Promise((res, rej) => {
      setTimeout(() => {
        res([
          {name: "Pinehurst Lake Conservation", comments: [{rating: 2}]},
          {name: "Pinehurst Lake Conservation", comments: [{rating: 5}]}
        ])
      }, 500)
    })
  }

  beforeEach(() => {
    component = shallow(
        <Campgrounds />
    )
  })

  it('Expect component to be defined', () => {
    expect(component).toBeDefined();
  })

  it('Before fetching data, no campgrounds shown', () => {
    expect(component.find('.CampgroundTile')).toHaveLength(0)
  })

  it('After fetching data, the right number of campgrounds are shown', () => {
    component.setState({
        campgrounds: [
          {id: 0, name: "Pinehurst Lake Conservation", comments: [{rating: 2}]},
          {id: 1, name: "Pinehurst Lake Conservation", comments: [{rating: 2}]}
        ]
    })

    expect(component.find('.CampgroundTile')).toHaveLength(2)
  })


  it('Campgrounds component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
})
