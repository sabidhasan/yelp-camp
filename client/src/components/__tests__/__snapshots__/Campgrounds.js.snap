// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Campgrounds /> Campgrounds component snapshot matches what is expected 1`] = `
<ul
  className="Campgrounds"
>
  <li
    className="Campgrounds__header flex-center"
  >
    <h1
      className="Campgrounds__header-title"
    >
      Some Awesome Campgrounds
    </h1>
    <i
      className="Campgrounds__random fas fa-sync-alt "
      onClick={[Function]}
    />
  </li>
</ul>
`;
